qiskit/transpiler/synthesis/aqc/aqc.py

```diff
@@ -10,30 +10,85 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 """A generic implementation of Approximate Quantum Compiler."""
-from typing import Optional
+from __future__ import annotations
+
+from functools import partial
+
+from collections.abc import Callable
+from typing import Protocol
 
 import numpy as np
+from scipy.optimize import OptimizeResult, minimize
 
-from qiskit.algorithms.optimizers import L_BFGS_B, Optimizer
+from qiskit.algorithms.optimizers import Optimizer
 from qiskit.quantum_info import Operator
+from qiskit.utils.deprecation import deprecate_arg
+
 from .approximate import ApproximateCircuit, ApproximatingObjective
 
 
+class Minimizer(Protocol):
+    """Callable Protocol for minimizer.
+
+    This interface is based on `SciPy's optimize module
+    <https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html>`__.
+
+     This protocol defines a callable taking the following parameters:
+
+         fun
+             The objective function to minimize.
+         x0
+             The initial point for the optimization.
+         jac
+             The gradient of the objective function.
+         bounds
+             Parameters bounds for the optimization. Note that these might not be supported
+             by all optimizers.
+
+     and which returns a SciPy minimization result object.
+    """
+
+    # pylint: disable=invalid-name
```

Comment:
This only applies to `x0`, right? It might be nicer to put the exception in that line 🙂